library/gen_x12_837.inc.php

```diff
@@ -1268,129 +1032,96 @@ function gen_x12_837($pid, $encounter, &$log, $encounter_claim = false)
-                $out .= "AMT" . // Approved amount per previous payer. Page 485.
-                "*AAE" .
-                "*" . sprintf('%.2f', $claim->cptCharges($prockey) - $payerpaid[2]) .
-                "~\n";
-                break;
-            }
-        }
+    if ($ndc) {
+        ++$edicount;
+    }
+    $out .= "LIN" . // Drug Identification. Page 500+ (Addendum pg 71).
+    "*" .         // Per addendum, LIN01 is not used.
+    "*" . "N4" .
+    "*" . $ndc .
+    "~\n";
 
-        // Loop 2410, Drug Information. Medicaid insurers seem to want this
-        // with HCPCS codes.
-        //
-        $ndc = $claim->cptNDCID($prockey);
-        if ($ndc) {
-            ++$edicount;
-            $out .= "LIN" . // Drug Identification. Page 500+ (Addendum pg 71).
-            "*" .         // Per addendum, LIN01 is not used.
-            "*N4" .
-            "*" . $ndc .
-            "~\n";
+    if (!preg_match('/^\d\d\d\d\d-\d\d\d\d-\d\d$/', $ndc, $tmp) && !preg_match('/^\d{11}$/', $ndc)) {
+        $log .= "*** NDC code '$ndc' has invalid format!\n";
+    }
 
-            if (!preg_match('/^\d\d\d\d\d-\d\d\d\d-\d\d$/', $ndc, $tmp) && !preg_match('/^\d{11}$/', $ndc)) {
-                $log .= "*** NDC code '$ndc' has invalid format!\n";
-            }
+    ++$edicount;
+    $out .= "CTP" . // Drug Pricing. Page 500+ (Addendum pg 74).
+    "*" .
+    "*" .
+    "*" .
+    "*" . $claim->cptNDCQuantity($prockey) .
+    "*" . $claim->cptNDCUOM($prockey) .
+    // Note: 5010 documents "ME" (Milligrams) as an additional unit of measure.
+    "~\n";
 
-            ++$edicount;
-            $tmpunits = $claim->cptNDCQuantity($prockey) * $claim->cptUnits($prockey);
-            if (!$tmpunits) {
-                $tmpunits = 1;
-            }
 
-            $out .= "CTP" . // Drug Pricing. Page 500+ (Addendum pg 74).
-            "*" .
-            "*" .
-            "*" . ($CMS_5010 ? '' : sprintf('%.2f', $claim->cptCharges($prockey) / $tmpunits)) .
-            "*" . $claim->cptNDCQuantity($prockey) .
-            "*" . $claim->cptNDCUOM($prockey) .
-            // Note: 5010 documents "ME" (Milligrams) as an additional unit of measure.
-            "~\n";
-        }
+    // Segment REF (Prescription or Compound Drug Association Number) omitted.
 
-        // Segment REF (Prescription or Compound Drug Association Number) omitted.
+    // Loop 2420A, Rendering Provider (service-specific).
+    // Used if the rendering provider for this service line is different
+    // from that in loop 2310B.
+    //
+    if ($claim->providerNPI() != $claim->providerNPI($prockey)) {
+        ++$edicount;
+        $out .= "NM1" .       // Loop 2310B Rendering Provider
+        "*" . "82" .
+        "*" . "1" .
+        "*" . $claim->providerLastName($prockey) .
+        "*" . $claim->providerFirstName($prockey) .
+        "*" . $claim->providerMiddleName($prockey) .
+        "*" .
+        "*";
+        if ($claim->providerNPI($prockey)) {
```

Comment:
more logging for invalid npi just below this i believe, this usually happens when first setting up practice and helps debugging